types/utils.d.ts

```diff
@@ -45,3 +45,37 @@ export type RequestHeadersDefault = unknown
 
 export type ContextConfigDefault = unknown
 export type ReplyDefault = unknown
+
+/**
+ * Helpers for determining the type of the response payload based on the code
+ */
+
+type WildCardKeys = {1: '1xx', 2:'2xx', 3:'3xx', 4:'4xx', 5:'5xx'};
+type ResponseCodes100 = 100 | 101 | 102 | 103 | 122;
+type ResponseCodes200 = 200 | 201 | 202 | 203 | 204 | 205 | 206 | 207 | 208 | 226;
+type ResponseCodes300 = 300 | 301 | 302 | 303 | 304 | 305 | 306 | 307 | 308;
+type ResponseCodes400 = 400 | 401 | 402 | 403 | 404 | 405 | 406 | 407 | 408 | 409 | 410 | 411 | 412 | 413 | 414 | 415 | 416 | 417 | 418 | 421 | 422 | 423 | 424 | 425 | 426 | 428 | 429 | 431 | 451;
+type ResponseCodes500 = 500 | 501 | 502 | 503 | 504 | 505 | 506 | 507 | 508 | 510 | 511;
+type ResponseCodes = ResponseCodes100 | ResponseCodes200 | ResponseCodes300 | ResponseCodes400 | ResponseCodes500;
+type ResponseKeys = WildCardKeys[keyof WildCardKeys] | ResponseCodes;
+
+export type StatusCodeReply = {
+  // eslint-disable-next-line no-unused-vars
+  [Key in ResponseKeys]?: unknown;
+};
+
+// weird TS quirk: https://stackoverflow.com/questions/58977876/generic-conditional-type-resolves-to-never-when-the-generic-type-is-set-to-never
+export type HttpCodesCovered<Key> = [Key] extends [never] ? number :
+  Key extends ResponseCodes ? Key :
+    Key extends `${infer X extends keyof WildCardKeys}xx` ?
+      X extends 1 ? ResponseCodes100 :
+        X extends 2 ? ResponseCodes200 :
+          X extends 3 ? ResponseCodes300 :
+            X extends 4 ? ResponseCodes400 :
+              X extends 5 ? ResponseCodes500 : never : number;
+
+export type CodeToReplyKey<Code extends ResponseKeys> = Code extends ResponseCodes100 ? '1xx' :
+  Code extends ResponseCodes200 ? '2xx' :
+    Code extends ResponseCodes300 ? '3xx' :
+      Code extends ResponseCodes400 ? '4xx' :
+        Code extends ResponseCodes500 ? '5xx' : Code;
```

Comment:
I would propose simplifying this to:

```suggestion
type StringDigit = '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9';

export type WildCardsToCode<Key> = Key extends number ? Key : number;
export type CodeToWildCard<Code extends number> = `${Code}` extends `${infer FirstDigit extends StringDigit}${infer _RemainingDigits}` ? `${FirstDigit}xx` : never;
```

Upsides:

* All response codes are allowed, the type system does not try to have a saying in if `102` is or isn't a valid HTTP response code
* Much shorter

Downsides:

* When `Reply` has a wildcard, then `WildCardsToCode<keyof Reply>` will become `number` rather than a list of all response codes that wildcard can resolve to. I think this is a good compromise until https://github.com/microsoft/TypeScript/issues/15480 at some point becomes a reality.